jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/internal/CompletedUpgradeRequest.java

```diff
@@ -0,0 +1,200 @@
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpScheme;
+import org.eclipse.jetty.http.HttpURI;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.util.Fields;
+import org.eclipse.jetty.websocket.api.ExtensionConfig;
+import org.eclipse.jetty.websocket.api.UpgradeRequest;
+import org.eclipse.jetty.websocket.common.JettyExtensionConfig;
+import org.eclipse.jetty.websocket.core.server.ServerUpgradeRequest;
+
+public class CompletedUpgradeRequest implements UpgradeRequest
+{
+    private final HttpFields _httpFields;
+    private final List<HttpCookie> _cookies;
+    private final List<ExtensionConfig> _extensions;
+    private final HttpURI _httpURI;
+    private final String _httpVersion;
+    private final String _method;
+    private final List<String> _subProtocols;
+    private final Map<String, List<String>> _queryParams;
+    private final boolean _secure;
+    private final String _protocolVersion;
+    private final Principal _userPrincipal;
+
+    public CompletedUpgradeRequest(ServerUpgradeRequest request)
+    {
+        _httpFields = request.getHeaders().asImmutable();
+        _httpURI = request.getHttpURI();
+        _method = request.getMethod();
+        _httpVersion = request.getConnectionMetaData().getHttpVersion().asString();
+        _subProtocols = request.getSubProtocols();
+        _secure = request.isSecure();
+        _protocolVersion = request.getProtocolVersion();
+
+        // Convert Jetty Cookies to java.net.HttpCookie.
+        List<HttpCookie> cookies = new ArrayList<>();
+        for (org.eclipse.jetty.http.HttpCookie cookie : Request.getCookies(request))
+        {
+            cookies.add(org.eclipse.jetty.http.HttpCookie.asJavaNetHttpCookie(cookie));
+        }
+        _cookies = Collections.unmodifiableList(cookies);
+
+        // Convert Core Extensions to Jetty API Extensions.
+        List<ExtensionConfig> extensions = new ArrayList<>();
+        for (org.eclipse.jetty.websocket.core.ExtensionConfig extension : request.getExtensions())
+        {
+            extensions.add(new JettyExtensionConfig(extension));
+        }
+        _extensions = Collections.unmodifiableList(extensions);
+
+        // Get the user principal from the request's authentication state.
+        Request.AuthenticationState authState = Request.getAuthenticationState(request);
+        _userPrincipal = authState != null ? authState.getUserPrincipal() : null;
+
+        // Extract query parameters from the request.
+        Map<String, List<String>> queryParams = new LinkedHashMap<>();
+        for (Fields.Field field : Request.extractQueryParameters(request))
+        {
+            queryParams.put(field.getName(), field.getValues());
+        }
+        _queryParams = Collections.unmodifiableMap(queryParams);
```

Comment:
It feels like querry params would almost certainly have already been extracted at this point.   Don't we store as an attribute or similar?    Hmm apparently  not... let's open another issue to investigate doing this in core (not just in servlet)

Reply:
Opened https://github.com/jetty/jetty.project/issues/13242